src/axom/primal/geometry/Quadrilateral.hpp

```diff
@@ -0,0 +1,264 @@
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+#ifndef AXOM_PRIMAL_QUADRILATERAL_HPP_
+#define AXOM_PRIMAL_QUADRILATERAL_HPP_
+
+#include "axom/config.hpp"
+#include "axom/core.hpp"
+#include "axom/slic/interface/slic.hpp"
+#include "axom/fmt.hpp"
+
+#include "axom/primal/constants.hpp"
+#include "axom/primal/geometry/Point.hpp"
+#include "axom/primal/geometry/Triangle.hpp"
+
+#include <cmath>
+#include <ostream>
+
+namespace axom
+{
+namespace primal
+{
+// Forward declare the templated classes and operator functions
+template <typename T, int NDIMS>
+class Quadrilateral;
+
+/**
+ * \brief Overloaded output operator for quadrilaterals
+ */
+template <typename T, int NDIMS>
+std::ostream& operator<<(std::ostream& os, const Quadrilateral<T, NDIMS>& quad);
+
+/*!
+ * \accelerated
+ * \class Quadrilateral
+ *
+ * \brief Represents a quadrilateral geometric shape defined by four points.
+ *
+ * \accelerated
+ *
+ * \tparam T the coordinate type, e.g., double, float, etc.
+ * \tparam NDIMS the number of dimensions
+ *
+ * There are four vertices in the quadrilateral, labeled A through D as the
+ * constructor's arguments.  They are accessible using the square-brackets
+ * operator, with A being index 0, B index 1, C index 2, and D index 3.
+ *
+ * Here's a diagram showing a square with the labeled vertices.
+ *
+ * \verbatim
+ *
+ * D +---------+ C       +y
+ *   |         |         
+ *   |         |          ^
+ *   |         |          |
+ *   |         |          |
+ * A +---------+ B        -----> +x
+ *
+ * \endverbatim
+ */
+template <typename T, int NDIMS>
+class Quadrilateral
+{
+public:
+  using PointType = Point<T, NDIMS>;
+  using TriangleType = Triangle<T, NDIMS>;
+
+  static constexpr int DIM = NDIMS;
+  static constexpr int NUM_QUAD_VERTS = 4;
+
+public:
+  /// \brief Default constructor. Creates a degenerate quadrilateral.
+  Quadrilateral() = default;
+
+  /*!
+   * \brief Custom Constructor. Creates a quadrilateral from the 4 points A, B, C, and D.
+   * \param [in] A point corresponding to vertex A of the quadrilateral.
+   * \param [in] B point corresponding to vertex B of the quadrilateral.
+   * \param [in] C point corresponding to vertex C of the quadrilateral.
+   * \param [in] D point corresponding to vertex D of the quadrilateral.
+   */
+  AXOM_HOST_DEVICE
+  Quadrilateral(const PointType& A,
+                const PointType& B,
+                const PointType& C,
+                const PointType& D)
+    : m_points {A, B, C, D}
+  { }
+
+  /*!
+   * \brief Quadrilateral constructor from an array of Points
+   *
+   * \param [in] pts An array containing at least 4 Points.
+   *
+   * \note It is the responsiblity of the caller to pass
+   *       an array with at least 4 Points
+   */
+  AXOM_HOST_DEVICE
+  explicit Quadrilateral(const PointType* pts)
+  {
+    for(int i = 0; i < NUM_QUAD_VERTS; i++)
+    {
+      m_points[i] = pts[i];
+    }
+  }
```

Comment:
Design: I think you added this because the other primitives have a similar constructor from a C-style array.
I'd suggest removing it, and if users want to use this, perhaps they can wrap their array in an `axom::ArrayView` and call the `Quadrilateral(ArrayView<Points>)` constructor.

Thoughts?

Reply:
Works for me. I'll make the change.